File: apps/macos/Tools/jpdrive/Sources/DriveKit/Ambient.swift
import AppKit
import Foundation

/// The state a driven run borrows from whoever is at the keyboard.
///
/// Which application is in front, and where the pointer is. Neither belongs to
/// the app under test: a synthesized gesture has to take both — mouse events go
/// to whatever is on top at a coordinate, and the ordering between applications
/// follows activation — and a run that takes them owes them back.
///
/// Deliberately not window geometry. A step that resizes a window did the thing
/// it was asked to do, and putting the window back would undo the effect under
/// test. What a run borrows is restored; what it was told to change is not.
///
/// Read and written separately rather than as one capture-and-restore pair, so a
/// caller composes what it needs and decides for itself when a restore is owed.
enum Ambient {
    /// The bundle identifier of the frontmost application.
    ///
    /// `nil` when there is none, or when it has no identifier — a process
    /// launched without a bundle has neither.
    static func frontmost() -> FrontmostReport {
        FrontmostReport(bundleID: NSWorkspace.shared.frontmostApplication?.bundleIdentifier)
    }

    /// Bring the application with `bundleID` back to the front.
    ///
    /// Through `NSWorkspace`, which asks the application to activate itself, so
    /// this needs no permission beyond launching one. Answers whether an
    /// application with that identifier was found to ask.
    static func activate(bundleID: String) -> FrontmostReport {
        guard
            let app = NSRunningApplication.runningApplications(withBundleIdentifier: bundleID)
                .first
        else {
            return FrontmostReport(bundleID: nil)
        }

        app.activate()
        return FrontmostReport(bundleID: bundleID)
    }

    /// Where the pointer is, in the coordinates a synthesized event uses.
    ///
    /// `NSEvent.mouseLocation` is bottom-left origin and screen coordinates are
    /// top-left, so the y is flipped here rather than at each call site. The
    /// height flipped against is the *main* screen's, which is what the window
    /// server measures global coordinates from.
    static func pointer() -> PointerReport {
        let location = NSEvent.mouseLocation
        let height = NSScreen.screens.first?.frame.height ?? 0

        return PointerReport(x: location.x, y: height - location.y)
    }

    /// Put the pointer back at `point`.
    ///
    /// Warped rather than moved: `CGWarpMouseCursorPosition` relocates the cursor
    /// without synthesizing motion, so nothing under it takes a hover, and no
    /// application sees a gesture it has to interpret.
    static func movePointer(to point: CGPoint) -> PointerReport {
        CGWarpMouseCursorPosition(point)
        return PointerReport(x: point.x, y: point.y)
    }
}

/// Which application is in front.
struct FrontmostReport: Encodable, Equatable {
    let bundleID: String?

    private enum CodingKeys: String, CodingKey {
        case bundleID = "bundle_id"
    }
}

/// Where the pointer is, in top-left-origin screen coordinates.
struct PointerReport: Encodable, Equatable {
    let x: CGFloat
    let y: CGFloat
}

File: .config/jp/tools/src/debug_app/ambient.rs
//! Borrowing and returning the state a driven run does not own.
//!
//! Which application is in front, and where the pointer is.
//! A step that synthesizes input has to take both: mouse events go to whatever
//! is on top at a coordinate, and the ordering between applications follows
//! activation.
//! Both belong to whoever is at the keyboard.
//!
//! Run-scoped, and it has to be: `jpdrive` runs one step per process, so
//! nothing on that side outlives a single step.
//! Restoring there would put focus back between every pair of steps and leave
//! the next one aiming at a window that is no longer in front.
//!
//! Deliberately not window geometry.
//! A step that resized a window did the thing it was asked to do, and putting
//! the window back would undo the effect the run was measuring.

use camino::Utf8Path;
use serde::Deserialize;

use crate::util::runner::ProcessRunner;

/// What a run borrowed, to be handed back when it ends.
///
/// Either field is `None` when the driver could not report it.
/// A restore skips what it does not know rather than guessing, because guessing
/// here moves the pointer of somebody who is using it.
#[derive(Debug, Default, PartialEq)]
pub(crate) struct Borrowed {
    frontmost: Option<String>,
    pointer: Option<(f64, f64)>,
}

#[derive(Deserialize)]
struct FrontmostReport {
    bundle_id: Option<String>,
}

#[derive(Deserialize)]
struct PointerReport {
    x: f64,
    y: f64,
}

/// Read what is about to be borrowed.
///
/// Never fails the run.
/// A driver that cannot report the frontmost application is a reason to leave
/// focus alone afterwards, not a reason to refuse to drive.
pub(crate) fn capture(bin: &Utf8Path, root: &Utf8Path, runner: &dyn ProcessRunner) -> Borrowed {
    Borrowed {
        frontmost: read::<FrontmostReport>(bin, &["frontmost"], root, runner)
            .and_then(|report| report.bundle_id),
        pointer: read::<PointerReport>(bin, &["pointer"], root, runner)
            .map(|report| (report.x, report.y)),
    }
}

/// Put back what was borrowed.
///
/// Focus first, then the pointer: activating an application does not move the
/// cursor, so the order only matters in that the pointer must not be placed and
/// then have an activation drag it elsewhere.
///
/// Silent about failure for the same reason as [`capture`]: this runs after the
/// work a caller asked for, and a complaint here would replace whatever the run
/// was reporting.
pub(crate) fn restore(
    borrowed: &Borrowed,
    bin: &Utf8Path,
    root: &Utf8Path,
    runner: &dyn ProcessRunner,
) {
    if let Some(bundle_id) = &borrowed.frontmost {
        let _restored = runner.run(bin.as_str(), &["frontmost", "--set", bundle_id], root);
    }

    if let Some((x, y)) = borrowed.pointer {
        let point = format!("{x},{y}");
        let _restored = runner.run(bin.as_str(), &["pointer", "--set", &point], root);
    }
}

/// Run one driver subcommand and read its JSON document.
fn read<T: for<'de> Deserialize<'de>>(
    bin: &Utf8Path,
    args: &[&str],
    root: &Utf8Path,
    runner: &dyn ProcessRunner,
) -> Option<T> {
    let output = runner.run(bin.as_str(), args, root).ok()?;
    serde_json::from_str(&output.stdout).ok()
}

#[cfg(all(test, unix))]
#[path = "ambient_tests.rs"]
mod tests;
